TRLevelControlTests/Base/TestBase.cs

```diff
@@ -0,0 +1,263 @@
+        writer.WriteLevelToFile(level, pathO);
+
+        byte[] b1 = File.ReadAllBytes(pathI);
+        byte[] b2 = File.ReadAllBytes(pathO);
+
+        CollectionAssert.AreEqual(b1, b2);
+    }
+
+    public static void ReadWriteTR3Level(string levelName)
+    {
+        TR3LevelReader reader = new();
+        TR3LevelWriter writer = new();
+
+        string pathI = GetReadPath(levelName, TRGameVersion.TR3);
+        string pathO = GetWritePath(levelName, TRGameVersion.TR3);
+
+        TR3Level level = reader.ReadLevel(pathI);
+        writer.WriteLevelToFile(level, pathO);
+
+        byte[] b1 = File.ReadAllBytes(pathI);
+        byte[] b2 = File.ReadAllBytes(pathO);
+
+        CollectionAssert.AreEqual(b1, b2);
+    }
+
+    public static void ReadWriteTR4Level(string levelName)
+    {
+        TR4LevelReader reader = new();
+        TR4LevelWriter writer = new();
+
+        string pathI = GetReadPath(levelName, TRGameVersion.TR4);
+        string pathO = GetWritePath(levelName, TRGameVersion.TR4);
+
+        // ZLib produces a slightly more optimal output than OG so we can't compare byte-for-byte
+        TR4Level level = reader.ReadLevel(pathI);
+        TR45LevelSummary originalSummary = new()
+        {
+            LevelChunkUncompressedSize = level.LevelDataChunk.UncompressedSize,
+            Tex32ChunkUncompressedSize = level.Texture32Chunk.UncompressedSize,
+            Tex16ChunkUncompressedSize = level.Texture16Chunk.UncompressedSize,
+            Tex32MChunkUncompressedSize = level.SkyAndFont32Chunk.UncompressedSize
+        };
+
+        writer.WriteLevelToFile(level, pathO);
+        // Read in again what we wrote out
+        TR4Level level2 = reader.ReadLevel(pathI);
+
+        // Verify - have we lost any data?
+        Assert.AreEqual(originalSummary.LevelChunkUncompressedSize, (uint)TRZlib.Decompress(level2.LevelDataChunk.CompressedChunk).Length);
+        Assert.AreEqual(originalSummary.Tex32ChunkUncompressedSize, (uint)TRZlib.Decompress(level2.Texture32Chunk.CompressedChunk).Length);
+        Assert.AreEqual(originalSummary.Tex16ChunkUncompressedSize, (uint)TRZlib.Decompress(level2.Texture16Chunk.CompressedChunk).Length);
+        Assert.AreEqual(originalSummary.Tex32MChunkUncompressedSize, (uint)TRZlib.Decompress(level2.SkyAndFont32Chunk.CompressedChunk).Length);
+
+        // Test compression against original
+        CollectionAssert.AreEqual(new byte[] { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }, level2.LevelDataChunk.Seperator);
+        CollectionAssert.AreEqual(TRZlib.Decompress(level.Texture32Chunk.CompressedChunk), TRZlib.Decompress(level2.Texture32Chunk.CompressedChunk));
+        CollectionAssert.AreEqual(TRZlib.Decompress(level.Texture16Chunk.CompressedChunk), TRZlib.Decompress(level2.Texture16Chunk.CompressedChunk));
+        CollectionAssert.AreEqual(TRZlib.Decompress(level.SkyAndFont32Chunk.CompressedChunk), TRZlib.Decompress(level2.SkyAndFont32Chunk.CompressedChunk));
```

Comment:
This specialization seems to be violating inversion of control rule, maybe it's better to create a comparator for levels of concrete engine?

Reply:
Yes, agreed - it's not ideal at the moment. This is a stopgap for the time being - we have an upgraded reader/writer library in the planning and it will come with improved tests where we will compare precisely the zipped/unzipped data (I have this in a separate project which we're going to be slowly migrating over to, as it's far too big a change to drop in one go). The TR4/5 IO is currently unused as it stands, so this will be a future task